### DISCUSSION TITLE
How to store virtual or evaluated columns in an extra file?

### DISCUSSION THREAD

**Discussion by Eisbrenner** [vaexio/vaex#1843] [Q&A] [answered]
Hi,

I like to extend a dataframe, which I have as a hdf5 file generated by vaex from a csv. However, I also have a pipeline where the input file is tracked and the pipeline is rerun if it changes. So I can't write additional columns into the same hdf5 file (modifying it in the process).

Further, some columns I would like to just compute (and only later decide what I do with them). Again, I don't want to write them into a new file, since I don't want the original hdf5 file to be changed.

I assume that I miss an obvious thing here, therefore I won't create a minimal example right now. If you think its needed, just let me know and I get to it.

A bit background,

1. I have a data file `data_all = vaex.open(path)` with about 5e9 rows.
2. To develop my script, I start out with a debug/dev line `data = data_all[slice(None, 1_000_000)]`.
3. Then I create two virtual columns with booleans in them (one of them beeing like this: `data["iswall"] = (data["i"] % 1.0 == 0) | (data["j"] % 1.0 == 0)`)
   + side note here, I like to compute some other things later, which I really only want to do once (for example `max( data.t[id] ) - min( data.t[id] )` but even worse `data["S"] = xarray_lookup_S(data.i, data.j, data.t)`)

Finally, I thought just write it onto disk. Next time I load and concatenate the two files and I'm good. So, in anther script the new columns will already be there and I can move past that stuff. However, that throws errors, saying that those additional columns are not writable.

```
data["iswall"].export_hdf5(
        data_dir.joinpath(
            *[f"{file_stem}.hdf5" if not debug else f"{file_stem}-debug.hdf5"]
        ),
    )
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/x_ezrei/data/output/DIANS/src/convert/create_iswall_column.py", line 77, in export_to_hdf5
    data.export_hdf5(
AttributeError: 'Expression' object has no attribute 'export_hdf5'
```

```
data["iswall"].evaluate().export_hdf5(
        data_dir.joinpath(
            *[f"{file_stem}.hdf5" if not debug else f"{file_stem}-debug.hdf5"]
        ),
    )
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/x_ezrei/data/output/DIANS/src/convert/create_iswall_column.py", line 77, in export_to_hdf5
    data.export_hdf5(
AttributeError: 'numpy.ndarray' object has no attribute 'export_hdf5'
```

Thanks
Ezra

**Comment by JovanVeljanoski** [comment, marked as answer]
The first error is obvious: you can't export an expression. You can export a dataframe with a single expression tho, so probably simples would be:

```python
data[["iswall"]].export_hdf5(...)
```

2nd error is also obvious:when you `evaluate` an expression, you basically get a numpy (or arrow) representation from it, an it now lives in memory outside of a Vaex dataframe. 

I hope this helps a bit.

**Reply by Eisbrenner** [reply]
Yes very much so! I slowly find my way around all of vaex and tabular data in general.

Thanks!
